business/proxy_status.go

```diff
@@ -0,0 +1,129 @@
+package business
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/models"
+)
+
+type ProxyStatus struct {
+	k8s kubernetes.ClientInterface
+}
+
+func (in *ProxyStatus) GetWorkloadProxyStatus(workloadName, namespace string, workloadStatus models.WorkloadStatus) models.WorkloadStatus {
+	proxyStatuses, err := in.k8s.GetProxyStatus()
+	if err != nil {
+		return workloadStatus
+	}
+
+	ps := getProxyStatusByName(workloadName, namespace, proxyStatuses)
+	if ps != nil {
+		workloadStatus.ProxyStatus = castProxyStatus(*ps)
+	}
+	return workloadStatus
+}
+
+func (in *ProxyStatus) GetWorkloadsProxyStatuses(namespace string, workloadStatuses []models.WorkloadStatus) []models.WorkloadStatus {
+	res := make([]models.WorkloadStatus, 0, len(workloadStatuses))
+
+	proxyStatuses, err := in.k8s.GetProxyStatus()
+	if err != nil {
+		return workloadStatuses
+	}
+
+	for _, ws := range workloadStatuses {
+		ps := getProxyStatusByName(ws.Name, namespace, proxyStatuses)
+		if ps != nil {
+			ws.ProxyStatus = castProxyStatus(*ps)
+		}
+		res = append(res, ws)
+	}
+
+	return res
+}
+
+func (in *ProxyStatus) GetNamespaceAppProxyStatus(namespace string, appsWorkloads models.NamespaceAppHealth) (models.NamespaceAppHealth, error) {
+	proxyStatuses, err := in.k8s.GetProxyStatus()
+	if err != nil {
+		return appsWorkloads, err
+	}
+
+	for app, workloadStatus := range appsWorkloads {
+		workloadStatuses := make([]models.WorkloadStatus, 0, len(workloadStatus.WorkloadStatuses))
+
+		for _, ws := range workloadStatus.WorkloadStatuses {
+			ps := getProxyStatusByName(ws.Name, namespace, proxyStatuses)
+			if ps != nil {
+				ws.ProxyStatus = castProxyStatus(*ps)
+			}
+			workloadStatuses = append(workloadStatuses, ws)
+		}
+		appsWorkloads[app].WorkloadStatuses = workloadStatuses
+	}
+
+	return appsWorkloads, nil
+}
+
+func (in *ProxyStatus) GetNamespaceWorkloadProxyStatus(namespace string, workloads models.NamespaceWorkloadHealth) (models.NamespaceWorkloadHealth, error) {
+	res := models.NamespaceWorkloadHealth{}
+
+	proxyStatuses, err := in.k8s.GetProxyStatus()
+	if err != nil {
+		return workloads, err
+	}
+
+	for wl, ws := range workloads {
+		ps := getProxyStatusByName(wl, namespace, proxyStatuses)
+		if ps != nil {
+			ws.WorkloadStatus.ProxyStatus = castProxyStatus(*ps)
+		}
+		res[wl] = ws
+	}
+
+	return res, nil
+}
+
+func getProxyStatusByName(name, namespace string, proxyStatus []*kubernetes.ProxyStatus) *kubernetes.ProxyStatus {
```

Comment:
Hmm, if there's several pods per workload, only the first match will be taken?

Reply:
You are right. In case of deployment which might have more than one workload. This will make the workload health display the proxy status differently.

Reply:
Mmm, note that a workload IS not a pod.
Workload is the controller + their pods.
So perhaps I wonder if the proxy status is something that should be tied to Pod instead and summarized into the Workload.

But I'm starting to review the whole feature, so take this comment with a pinch of salt.